### PR TITLE
Improve Rustdoc UI for scraped examples with multiline arguments, fix overflow in line numbers

### PR DESCRIPTION
**Pull request by willcrichton**
This PR improves a few aspects of the scrape examples feature in Rustdoc.
* Only function names and not the full call expression are highlighted.
* For call-sites with multiline arguments, the minimized code viewer will scroll to the top of the call-site rather than the middle if the argument is larger than the viewer size, ensuring that the function name is visible.
* This fixes an issue where the line numbers column had a visible x-scroll bar.

r? @GuillaumeGomez

### CODE REVIEW

File: src/librustdoc/scrape_examples.rs
Comment (camelid) [inline src/librustdoc/scrape_examples.rs:177]: I'm pretty sure you can just use the first field of `MethodCall`. It's the path segment for the method name. Also, this Span-manipulating code will break with macros.

Comment (willcrichton) [reply src/librustdoc/scrape_examples.rs:177]: Ah, I was confused b/c `PathSegment` doesn't have a span. But I see that `PathSegment::ident` has a span. Should I just use that, and ignore stuff like type parameters?

Comment (camelid) [reply src/librustdoc/scrape_examples.rs:177]: I think ignoring type parameters sounds good. So, yeah, just use the ident's span.

File: src/librustdoc/html/render/mod.rs
Comment (camelid) [inline src/librustdoc/html/render/mod.rs:2587]: This will make the link have the right release channel.

```suggestion
             <a class=\"scrape-help\" href=\"{doc_prefix}/rustdoc/scraped-examples.html\" target=\"_blank\">?</a>\
          </h5>",
        id = id,
        doc_prefix = DOC_RUST_LANG_ORG_CHANNEL,
    );
```

File: src/doc/rustdoc/src/scraped-examples.md
Comment (camelid) [inline src/doc/rustdoc/src/scraped-examples.md:1]: Can you move the part about how this feature is unstable up to the top of the page? Otherwise, the first few paragraphs make it look like this feature is enabled-by-default.

**Comment by camelid** [comment]
Please move the `--test` commit to a new PR. It's a separate issue from improving the UI.

**Comment by camelid** [comment]
I think that change would also require an FCP and more discussion.

**Comment by willcrichton** [comment]
Ok, I pulled that change into #93497.

File: src/librustdoc/html/render/mod.rs
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/mod.rs:2581]: I think this is way too much:
 * It's pretty obvious what this is.
 * It'll increase the page size once more for a very minor gain.

Comment (willcrichton) [reply src/librustdoc/html/render/mod.rs:2581]: cc @jsha and @camelid who have thoughts on this

File: src/librustdoc/html/render/mod.rs
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/mod.rs:2754]: Just realized: could you use `<details>` and `<summary>` to do this please? We use it for doc blocks and it has a few advantages like: no JS. :D

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/mod.rs:2754]: That would also simplify your CSS since we already have the CSS for it.

Comment (willcrichton) [reply src/librustdoc/html/render/mod.rs:2754]: Can you explain a bit more what you mean? This UI is already using `<details>` and `<summary>`, so I'm not sure what you would like changed. The added line is text that shows up once you expand the `details`.

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/mod.rs:2754]: Ah I see! Sorry I missed it. Then nevermind. :)

File: src/librustdoc/html/render/mod.rs
Comment (camelid) [reply src/librustdoc/html/render/mod.rs:2581]: I really don't think we should be worrying about a small increase in page size like this. And I don't think it's that obvious what it is. For you, it may be obvious, since you've been immersed in this feature throughout its development, but most users have never even heard of it.

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/mod.rs:2581]: Even if they don't know about the feature, the text itself is pretty explicit: "Examples found in...". The rustdoc book exists and will include a chapter about this but I don't think there is enough interest for it.

Also, it's definitely not a small increase in page size. Especially on big crates with a lot of methods.

Comment (camelid) [reply src/librustdoc/html/render/mod.rs:2581]: > The rustdoc book exists and will include a chapter about this but I don't think there is enough interest for it.

I'm not sure what you mean about interest.

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/mod.rs:2581]: Let me reword: I tried to say that since the rustdoc book exists and the feature itself is quite explicit, there is no point to include a link every time linking to the book chapter directly.

Now that I think about it, linking to the rustdoc book directly is also problematic for stable/beta/nightly links because the name of the chapter or its location could change whereas the link will stay around at a previous location

Comment (camelid) [reply src/librustdoc/html/render/mod.rs:2581]: > Now that I think about it, linking to the rustdoc book directly is also problematic for stable/beta/nightly links because the name of the chapter or its location could change whereas the link will stay around at a previous location

The code explicitly uses the correct channel, so that shouldn't be an issue in most cases. Since docs.rs uses nightly, it might not be solved there (since only the latest nightly is kept), but I don't think it's of concern.

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/mod.rs:2581]: Let me show an example: you are using the current stable, 1.58, so the link is pointing to `doc.rust-lang.org/book/stable/this-chapter` (something along the line, the `/stable/` is the important part). Then a few months later, after a few updates in the books, the link changed but the URL still points to the previous one which is now invalid. Linking to external resources is generally a terrible idea.

If we really want to have some extra explanation, we need to find another way like having a popup with short explanation and saying to take a look at the rustdoc book, or something else which doesn't link directly to external resources.

Comment (camelid) [reply src/librustdoc/html/render/mod.rs:2581]: No, the link will be `/1.58.0/`, not `/stable/`.

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/mod.rs:2581]: Not on docs.rs, which is literally the biggest website containing all crates documentation.

Comment (camelid) [reply src/librustdoc/html/render/mod.rs:2581]: Well, on docs.rs, it will be `/nightly/`, but I see your point. I still think it's unlikely for the docs to move much. If it becomes an issue, we could always change it.

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/mod.rs:2581]: I think it's big enough of a potential problem to not be added. And there is also the size increase issue too.

Comment (camelid) [reply src/librustdoc/html/render/mod.rs:2581]: I still think clearly communicating about features like this is a good idea, but I don't want to argue about it; it's not important enough.

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/mod.rs:2581]: I agree on the communication part. I don't agree on how it's currently done. Sorry for nitpicking on this but I think it's really important to have something working on all situations...

Comment (willcrichton) [reply src/librustdoc/html/render/mod.rs:2581]: @GuillaumeGomez if it's primarily page size you're worried about, I could always move the rendering logic to Javascript. I could append the `<a>` after the page loads.

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/mod.rs:2581]: The page size is easy to fix indeed (I'd prefer to have it done on click only instead of generating it for all items, but that's another debate). My main worry is really about the link target.

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/mod.rs:2581]: @willcrichton Like I said, instead of a link, maybe a small popup like the help menu which would say something like: "this was generated by the scrape-examples feature. You can find more information about it in the rustdoc book" where `rustdoc book` is a link to the rustdoc book (but not to the chapter to prevent this current issue).

The idea behind this is to name the feature so people know what to look for and provide them a link where they can find the information without worrying about the link being broken.

Does it sound good to you? (and to you too @camelid and @jsha?)

Comment (jsha) [reply src/librustdoc/html/render/mod.rs:2581]: @GuillaumeGomez I think you're proposing to have a tooltip with a clickable link inside it. That's a bad UI because it creates [hover tunnels](https://uxmovement.com/navigation/why-hover-menus-do-users-more-harm-than-good/).

Also, the problem you're trying to address is: the destination of the link might 404 in the future. You'd like to solve it by: making the link point to the whole book rather than a chapter. Whether we have a popup or a direct link is orthogonal to that question.

I'd like to solve that problem by: If the name of the chapter changes, put in a redirect. If mdbook doesn't support redirects, we should add it. They'd probably have to be JS redirects since mdbook just emits HTML, not server configs. But that's fine.

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/mod.rs:2581]: I'm really not a big fan of that idea but that sounds like a good compromise.

Comment (jsha) [reply src/librustdoc/html/render/mod.rs:2581]: Ideally mdbook would also emit some sort of redirect map that can be consumed by server configs, so aware servers (like doc.rust-lang.org) can do a proper 302. But at a minimum it would need to support a JS redirect, in case a book is uploaded to an unconfigured server, or if it is being served locally.

Comment (willcrichton) [reply src/librustdoc/html/render/mod.rs:2581]: So is the resolution: we'll leave the code as-is, and deal with adding a redirect if the issue arises?

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/mod.rs:2581]: I'd prefer to have mdbook update done before merging this because TODOs tend to remain as such. ^^'

Also, please generate this link using JS as you suggested. To get the current rustdoc version, you can use `getVar("rustdoc-version")` in the JS (you can see the result at the bottom of the help popup).

Comment (jsha) [reply src/librustdoc/html/render/mod.rs:2581]: @GuillaumeGomez I don't think it makes sense to hold up this feature on an mdbook feature, on the purely hypothetical idea that the chapter name will change. Note that this isn't the sort of TODO that will linger indefinitely. _If_ we rename the chapter, that will be the forcing function to implement redirects in mdbook.

How about this: In the Markdown source for mdbook, we add a comment: "Don't rename this chapter. If you do, make sure there's a redirect from the old name, so links from old documentation will continue to work."

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/mod.rs:2581]: I have very cold feet on this. Linking to external resources on a stable feature is forever. If anything changes (from the domain name to the chapter name, pretty much anything), it's a dead link and we can't fix it. It's terrible user-wise.

Comment (jsha) [reply src/librustdoc/html/render/mod.rs:2581]: @GuillaumeGomez how about another possibility: What if we create a small, dedicated documentation page for the scrape examples and emit it with each documentation build? So we could link to scraped-examples.html within the current doc directory and be fully contained?

Comment (camelid) [reply src/librustdoc/html/render/mod.rs:2581]: I'm pretty sure mdBook already has a redirection feature. It's used by the rustc-dev-guide.

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/mod.rs:2581]: @jsha: I think I prefer your solution a lot more. Like that, whether you have an internet connection or not, it'll work and we don't have to worry about the target being moved.

Comment (willcrichton) [reply src/librustdoc/html/render/mod.rs:2581]: If we're going to have an explainer in the docs that's distinct from the chapter in the rustdoc book, it should probably just be an overlay / popup like the current help menu. It wouldn't be more than a couple sentences probably.

Comment (willcrichton) [reply src/librustdoc/html/render/mod.rs:2665]: @jsha @GuillaumeGomez can y'all let me know what your final preference is for the scrape examples help message? I'd like to implement it this weekend.

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/mod.rs:2665]: Writing a page and linking to it based on what @jsha suggested.

Comment (willcrichton) [reply src/librustdoc/html/render/mod.rs:2665]: Is it ok if I implement it as a popup though? I really think that would be both a better user experience and simpler for me to implement.

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/mod.rs:2665]: Simpler I'm not sure. Adding an HTML page should be quite easy. Then you just have to link to it. Adding more JS for this doesn't seem the best solution but maybe I'm wrong?

**Comment by willcrichton** [comment]
Ok I've added a built-in help page. Here's what it looks like: https://willcrichton.net/example-analyzer/scrape-examples-help.html

A question: is it important to have both this page and an entry in the rustdoc book? What content should go in one or the other?

**Comment by willcrichton** [comment]
@jsha let me know your preference and I can get this PR wrapped up this weekend!

**Comment by jsha** [comment]
> A question: is it important to have both this page and an entry in the rustdoc book? What content should go in one or the other?

I don't think it's important to have both. If we have this page, IMO we don't need one in the book too.

**Comment by willcrichton** [comment]
Alright, well if it's good as is, are there any last requests on this PR? @jsha @GuillaumeGomez @camelid

**Comment by camelid** [comment]
The effects of this change as listed in the PR description sound good to me; I'll leave reviewing the code to Guillaume.

File: src/librustdoc/html/static/scrape-examples-help.md
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/scrape-examples-help.md:34]: Can you add a new section at the end providing a link to the rustdoc book chapter talking about this feature please? Something like:

```txt
## Official tutorial

If you want more information about this feature, take a look at the [rustdoc book chapter](link here).
```

**Comment by GuillaumeGomez** [comment]
Apart from my comment, looks all good to me!

File: src/librustdoc/html/static/scrape-examples-help.md
Comment (willcrichton) [reply src/librustdoc/html/static/scrape-examples-help.md:34]: Just added a commit for this. It looks like:

Screen Shot 2022-04-12 at 11 04 39 AM

**Comment by willcrichton** [comment]
Note: I also just added a commit which removes the remaining `unwrap`s in `scrape_examples.rs` to harden the scraping phase. I would like to make sure that once this feature rolls out, people encounter issues as "my example isn't being scraped" rather than "rustdoc is panicking".

File: src/librustdoc/html/render/mod.rs
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/mod.rs:475]: Nice!

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
All good for me. Then let's go!

@bors: r+

**Pull request merged**
